Perform plugin mounts in the runtime

cpuguy83 opened the pull request:
- For metrics plugins, mount the metrics socket using the runtime spec rather than on the host. This prevents mount leakages of the socket which can lead to failed plugin removal. This made it necessary to expose an interface to subsystems that interact with plugins to be able to modify the runtime spec.
- Make the "propagated" mount for plugins happen in the runtime spec. This also prevents mount leakages from the plugin rootfs
- Have volume plugins request a scoped path from the plugin (object) rather than assuming a hard-coded base path. This is required due to changes to using the runtime spec to handle the "propagated mount" for the plugin.
- Revert 0c2821d6f2de692d105e50a399daa65169697cca -- private /var/lib/docker/plugins is no longer needed and causes issues for other systems.
- Do not use "lazy" unmounts in recursive unmount, which can cause issues with submounts that exist in running containers, additionally makes sure plugin shutdown uses recursive unmounts.

cpuguy83 committed:
Use runtime spec modifier for metrics plugin hook

Currently the metrics plugin uses a really hackish host mount with
propagated mounts to get the metrics socket into a plugin after the
plugin is alreay running.
This approach ends up leaking mounts which requires setting the plugin
manager root to private, which causes some other issues.

With this change, plugin subsystems can register a set of modifiers to
apply to the plugin's runtime spec before the plugin is ever started.
This will help to generalize some of the customization work that needs
to happen for various plugin subsystems (and future ones).

Specifically it lets the metrics plugin subsystem append a mount to the
runtime spec to mount the metrics socket in the plugin's mount namespace
rather than the host's and prevetns any leaking due to this mount.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 committed:
Revert "Make plugins dir private."

This reverts commit 0c2821d6f2de692d105e50a399daa65169697cca.

Due to other changes this is no longer needed and resolves some other
issues with plugins.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 committed:
Plugins perform propagated mount in runtime spec

Setting up the mounts on the host increases chances of mount leakage and
makes for more cleanup after the plugin has stopped.
With this change all mounts for the plugin are performed by the
container runtime and automatically cleaned up when the container exits.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 committed:
Ensure plugin returns correctly scoped paths

Before this change, volume management was relying on the fact that
everything the plugin mounts is visible on the host within the plugin's
rootfs. In practice this caused some issues with mount leaks, so we
changed the behavior such that mounts are not visible on the plugin's
rootfs, but available outside of it, which breaks volume management.

To fix the issue, allow the plugin to scope the path correctly rather
than assuming that everything is visible in `p.Rootfs`.
In practice this is just scoping the `PropagatedMount` paths to the
correct host path.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "no_private_mount_for_plugins" git@github.com:cpuguy83/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842354329048
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

cpuguy83 committed:
Make sure plugin mounts are cleaned up

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

thaJeztah commented on volume/drivers/adapter.go:
```diff
@@ -16,7 +15,7 @@ var (
 type volumeDriverAdapter struct {
 	name         string
-	baseHostPath string
+	scopePath    func(s string) string
```
Should this be `scopedPath` (with a `d`)?

cpuguy83 replied:
I did this as is because it's an action e.g. `scopPath("/foo")`, otherwise it would be `getScopedPath("/foo")`

thaJeztah replied:
Oh! I see :blush:

thaJeztah commented on volume/drivers/adapter.go:
```diff
@@ -30,24 +29,17 @@ func (a *volumeDriverAdapter) Create(name string, opts map[string]string) (volum
+		proxy:      a.proxy,
+		name:       name,
+		driverName: a.name,
+		scopePath:  a.scopePath,
```
Should this be `scopedPath` (with a `d`)?

thaJeztah replied:
(same for other instances below)

thaJeztah commented:
Needs a minor rebase

ping @anusha-ragunathan @kolyshkin PTAL

cpuguy83 commented:
Rebased and fixed test issues.

thaJeztah reviewed (changes requested):
left one nit, but overall looks good

would like a review from @anusha-ragunathan, @kolyshkin though

cpuguy83 commented on plugin/manager_linux.go:
```diff
@@ -39,20 +39,16 @@ func (pm *Manager) enable(p *v2.Plugin, c *controller, force bool) (err error) {
 		}
 
-		if err = mount.MakeRShared(propRoot); err != nil {
+		if err := mount.MakeRShared(propRoot); err != nil {
```
Done

cpuguy83 commented:
Ping

anusha-ragunathan commented on integration-cli/docker_cli_daemon_plugins_test.go:
```diff
@@ -236,22 +221,6 @@ func (s *DockerDaemonSuite) TestVolumePlugin(c *check.C) {
-	}
-	mountPoint = strings.TrimSpace(mountPoint)
-
-	out, err = s.d.Cmd("run", "--rm", "-v", volName+":"+destDir, "busybox", "touch", destDir+destFile)
```
Can we add a way to test using the volume in a container?

cpuguy83 replied:
I'm not really sure what we're testing here.
It looks like we're just testing the the volume data is created in the plugin root, but this isn't really a valid test.
The only thing I can really think to do is to test if using the same volume name yields the same data, which is tested elsewhere.

Do you have an idea?

anusha-ragunathan replied:
Create data in one container and use it in another? 
We are also missing `propagated-mount`  related tests.

cpuguy83 replied:
We can do that, but then we're testing the plugin and not docker.

anusha-ragunathan replied:
The way I see it is, we are using a mock plugin to ensure docker is offering the said behavior.

cpuguy83 replied:
Updated the test

anusha-ragunathan commented on plugin/manager.go:
```diff
@@ -156,6 +157,10 @@ func (pm *Manager) HandleExitEvent(id string) error {
 		pm.enable(p, c, true)
 	}
+
+	if err := mount.RecursiveUnmount(filepath.Join(pm.config.Root, id)); err != nil {
```
For the enable case (line 158), is it okay to umount here?

cpuguy83 replied:
Good catch. Updated.

cpuguy83 commented:
This was updated and is all green.

anusha-ragunathan commented on daemon/metrics_unix.go:
```diff
@@ -34,52 +34,22 @@ func (daemon *Daemon) listenMetricsSock() (string, error) {
-	getter.Handle(metricsPluginType, func(name string, client *plugins.Client) {
+func registerMetricsPluginCallback(store *plugin.Store, sockPath string) {
+	store.RegisterRuntimeOpt(metricsPluginType, func(s *specs.Spec) {
+		f := plugin.WithSpecMounts([]specs.Mount{
```
Please update https://github.com/docker/cli/blob/master/docs/extend/config.md with this new option.

cpuguy83 replied:
I don't think there's a new option here?

anusha-ragunathan replied:
Sorry, not a user facing option. More of a methodology of how a plugin author can customize. But thinking more, its not necessary atm.

anusha-ragunathan commented on plugin/store.go:
```diff
@@ -204,6 +232,15 @@ func (ps *Store) Handle(capability string, callback func(string, *plugins.Client
 }
 
+// RegisterRuntimeOpt stores a list of SpecOpts for the provided capability.
+// These options are applied to the runtime spec before a plugin is started for the specified capability.
```
Can we use this method for the customizations we do for volume and graphdriver plugins?

cpuguy83 replied:
We can possibly, but I'd like to do it separately.
The changes I did include here are pretty numerous just because of attempting to achieve my end goal for this PR (do not use a private mount on the plugin dir), all these things had to happen first.

anusha-ragunathan commented on plugin/manager.go:
```diff
@@ -151,16 +150,6 @@ func (pm *Manager) HandleExitEvent(id string) error {
-		if err := mount.Unmount(p.PropagatedMount); err != nil {
-			logrus.Warnf("Could not unmount %s: %v", p.PropagatedMount, err)
-		}
-		propRoot := filepath.Join(filepath.Dir(p.Rootfs), "propagated-mount")
```
Should we continue to unmount `propRoot` here?

cpuguy83 replied:
The recursive unmount should take care of this.

anusha-ragunathan commented:
Were we able to test some volume and network plugins with this change?

anusha-ragunathan commented:
This version of the PR looks mostly good to me.

thaJeztah commented:
ping @cpuguy83 I see some minor nits from Anusha; anything else to address?

cpuguy83 committed:
Cleanup volume plugin test with bad assumptions

Test made some bad assumptions about on-disk state of volume data.
This updates the test to only test based on what the volume API is
designed to provide.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
I have tested this works with rexray.
Everything works without a hitch, mounts are as expected on the host.

anusha-ragunathan commented:
LGTM

vdemeester reviewed (approved):
LGTM :cactus:

thaJeztah reviewed (approved):
LGTM
